src/output/naive.rs

```diff
@@ -8,258 +8,214 @@
+    let all_points: BTreeSet<Point> = all_facts
+        .cfg_edge
+        .iter()
+        .map(|&(p, _)| p)
+        .chain(all_facts.cfg_edge.iter().map(|&(_, q)| q))
+        .collect();
+
+    for &r in &all_facts.universal_region {
+        for &p in &all_points {
+            all_facts.region_live_at.push((r, p));
+        }
+    }
+
+    let mut result = Output::new(dump_enabled);
+
+    let subset_start = Instant::now();
+
+    // compute the subsets rules, but indexed ((Region, Point), Region) for the next iteration
+    let subset = {
+        // Create a new iteration context, ...
+        let mut iteration1 = Iteration::new();
+
+        // .. some variables, ..
+        let subset = iteration1.variable::<(Region, Region, Point)>("subset");
+
+        // different indices for `subset`.
+        let subset_r1p = iteration1.variable::<((Region, Point), Region)>("subset_r1p");
+        let subset_r2p = iteration1.variable::<((Region, Point), Region)>("subset_r2p");
+        let subset_p = iteration1.variable::<(Point, (Region, Region))>("subset_p");
+
+        // temporaries as we perform a multi-way join.
+        let subset_1 = iteration1.variable::<((Region, Point), Region)>("subset_1");
+        let subset_2 = iteration1.variable::<((Region, Point), Region)>("subset_2");
+
+        let region_live_at = iteration1.variable::<((Region, Point), ())>("region_live_at"); // redundantly computed index
+        let cfg_edge_p = iteration1.variable::<(Point, Point)>("cfg_edge_p"); // redundantly computed index
+
+        // load initial facts.
+        subset.insert(all_facts.outlives.into());
+        region_live_at.insert(Relation::from(all_facts.region_live_at.iter().map(|&(r, p)| ((r, p), ()))));
+        cfg_edge_p.insert(all_facts.cfg_edge.clone().into());
+
+        // .. and then start iterating rules!
+        while iteration1.changed() {
+
+            // remap fields to re-index by keys.
+            subset_r1p.from_map(&subset, |&(r1,r2,p)| ((r1,p),r2));
+            subset_r2p.from_map(&subset, |&(r1,r2,p)| ((r2,p),r1));
+            subset_p.from_map(&subset, |&(r1,r2,p)| (p,(r1,r2)));
+
+            // subset(R1, R2, P) :- outlives(R1, R2, P).
+            // Already loaded; outlives is static.
+
+            // subset(R1, R3, P) :-
+            //   subset(R1, R2, P),
+            //   subset(R2, R3, P).
+            subset.from_join(&subset_r2p, &subset_r1p, |&(_r2,p),&r1,&r3| (r1,r3,p));
+
+            // subset(R1, R2, Q) :-
+            //   subset(R1, R2, P),
+            //   cfg_edge(P, Q),
+            //   region_live_at(R1, Q),
+            //   region_live_at(R2, Q).
+
+            subset_1.from_join(&subset_p, &cfg_edge_p, |&_p, &(r1,r2), &q| ((r1,q),r2));
+            subset_2.from_join(&subset_1, &region_live_at, |&(r1,q),&r2,&()| ((r2,q),r1));
+            subset.from_join(&subset_2, &region_live_at, |&(r2,q),&r1,&()| (r1,r2,q));
+        }
+
+        subset_r1p.complete()
```

Comment:
Just to check my understanding here: it seems to me that the `subset` variable is not needed. We *could* just use `subset_r1p` instead? (That is, nothing directly joins from `subset`; the only use of it is to be mapped into different arrangements, and we could do those maps from `subset_r1p` just as easily, right?)

(cc @frankmcsherry)

Reply:
Answer: yes, I checked elsewhere, and even got a slight win from doing so (in datafrog-opt).